### PR TITLE
ranger: fix the range construction behavior when the column's type is `YEAR`

### LINKED ISSUE #23508: “ Unexpected ExprType Float64 and EvalType Int” while using cluster index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
CREATE TABLE `PK_S_MULTI_42_2` (  `COL1` year(4) NOT NULL,  `COL2` year(4) NOT NULL,  `COL3` year(4) DEFAULT NULL,  PRIMARY KEY (`COL1`,`COL2`) clustered);
select * from PK_S_MULTI_42 where col2 between 2154 and 2154 AND col3 > 2153;
```

### 2. What did you expect to see? (Required)
execute  successfully without error 
### 3. What did you see instead (Required)
```sql
mysql > CREATE TABLE `PK_S_MULTI_42_2` (  `COL1` year(4) NOT NULL,  `COL2` year(4) NOT NULL,  `COL3` year(4) DEFAULT NULL,  PRIMARY KEY (`COL1`,`COL2`) clustered);
Query OK, 0 rows affected
Time: 0.282s
mysql > select * from PK_S_MULTI_42_2 where col2 between 2154 and 2154 AND col3 > 2153;
(1105, 'other error: [components/tidb_query_expr/src/types/expr_builder.rs:357]: Unexpected ExprType Float64 and EvalType Int')

mysql > CREATE TABLE `PK_S_MULTI_42_1` (  `COL1` year(4) NOT NULL,  `COL2` year(4) NOT NULL,  `COL3` year(4) DEFAULT NULL,  PRIMARY KEY (`COL1`,`COL2`) nonclustered);
Query OK, 0 rows affected
Time: 0.305s
mysql > select * from PK_S_MULTI_42_1 where col2 between 2154 and 2154 AND col3 > 2153;
+------+------+------+
| COL1 | COL2 | COL3 |
+------+------+------+
```
### 4. What is your TiDB version? (Required)
master: fd706ab76bd09ac859

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #23508 

Problem Summary:

The type of year should be int, not double.

### What is changed and how it works?

Use ConvertToMysqlYear() to adjust the year constant, and handle some corner cases.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

- Manual test (add detailed scripts or steps below)

Start a tidb cluster.
CREATE TABLE `PK_S_MULTI_42_2` (  `COL1` year(4) NOT NULL,  `COL2` year(4) NOT NULL,  `COL3` year(4) DEFAULT NULL,  PRIMARY KEY (`COL1`,`COL2`) clustered);
mysql> select * from PK_S_MULTI_42_2 where col2 between 2154 and 2154 AND col3 > 2153;
Empty set (0.00 sec)

Side effects

### Release note 

- fix wrong year type

### CODE REVIEW

**Comment by dyzsr** [comment]
Can someone help me /lgtm?

File: util/ranger/points.go [hunk 1]
```diff
@@ -221,31 +228,49 @@ func (r *builder) buildFormBinOp(expr *expression.ScalarFunction) []*point {
...
 			value.SetString(value.GetString(), col.RetType.Collate)
 		}
 		if col.GetType().Tp == mysql.TypeYear {
-			*value, err = types.ConvertDatumToFloatYear(r.sc, *value)
+			preValue, err1 := value.ToInt64(r.sc)
+			if err1 != nil {
+				return err1
+			}
+			*value, err = value.ConvertToMysqlYear(r.sc, col.RetType)
+			if errors.ErrorEqual(err, types.ErrInvalidYear) {
```
Comment (winoros) [inline util/ranger/points.go:244]: Could we add some comments here?

File: util/ranger/points.go [hunk 2]
```diff
@@ -472,10 +499,9 @@ func (r *builder) buildFromIn(expr *expression.ScalarFunction) ([]*point, bool)
 			dt.SetString(dt.GetString(), colCollate)
 		}
 		if expr.GetArgs()[0].GetType().Tp == mysql.TypeYear {
-			dt, err = types.ConvertDatumToFloatYear(r.sc, dt)
+			dt, err = dt.ConvertToMysqlYear(r.sc, expr.GetArgs()[0].GetType())
 			if err != nil {
-				r.err = ErrUnsupportedType.GenWithStack("expr:%v is not converted to year", e)
-				return getFullRange(), hasNull
+				continue
```
Comment (winoros) [inline util/ranger/points.go:504]: Add comments here also

**Comment by lysu** [comment]
do we need fix this in 5.0 GA? @wjhuang2016 @winoros  - -

**Comment by wjhuang2016** [comment]
> do we need fix this in 5.0 GA? @wjhuang2016 @winoros - -

Sure , why not?

**Review by winoros — APPROVED** [review]

**Review by qw4990 — COMMENTED** [review]
@wjhuang2016  Please resolve conflicts at your convenience. And should we add some new and separated test cases for this PR?

**Comment by qw4990** [comment]
I found that in the latest Master branch with commit ID `dbd5a3622a537c4fcb8324636c764caa0f5ac0fb` the problem mentioned in https://github.com/pingcap/tidb/issues/23508 cannot be reproduced:
```
mysql> CREATE TABLE `PK_S_MULTI_42_2` (  `COL1` year(4) NOT NULL,  `COL2` year(4) NOT NULL,  `COL3` year(4) DEFAULT NULL,  PRIMARY KEY (`COL1`,`COL2`) clustered);
Query OK, 0 rows affected (0.00 sec)

mysql> select * from PK_S_MULTI_42_2 where col2 between 2154 and 2154 AND col3 > 2153;;
Empty set (0.00 sec)
```
PTAL @wjhuang2016

**Review by qw4990 — APPROVED** [review]

**Pull request merged**